cpp/src/arrow/array/builder_binary.h

```diff
@@ -189,6 +189,18 @@ class ARROW_EXPORT FixedSizeBinaryBuilder : public ArrayBuilder {
     return byte_builder_.Append(value, byte_width_);
   }
 
+  void UnsafeAppend(const uint8_t* value) {
+    UnsafeAppendToBitmap(true);
+    byte_builder_.UnsafeAppend(value, byte_width_);
+  }
+
+  void UnsafeAppend(util::string_view value) {
+#ifndef NDEBUG
```

Comment:
The UnsafeAppend methods return `void` rather than `Status`, so they would need to use `DCHECK` in any case. I'll refactor `CheckValueSize()` to be inline and empty for `NDEBUG`, which will reduce `#ifdef` proliferation but not performance

Reply:
The UnsafeAppend methods return void rather than Status, so they would need to use DCHECK in any case. I'll refactor CheckValueSize() to be inline and empty for NDEBUG, which will reduce #ifdef proliferation but not performance

Reply:
The issue is that DCHECK can't be in headers, if you can make it so, do it :)

Reply:
https://github.com/apache/arrow/pull/3880/files#diff-011119a62ed7b6dcdc29e9e46674ee94R254